### PR TITLE
SYNC-356 - initial unit test that demonstrates failed sync with Patie…

### PR DESCRIPTION
**Pull request by mseaton**
…nt Program Attributes

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/sync/SyncUtil.java [hunk 1]
```diff
@@ -214,19 +213,33 @@ public static NodeList getChildNodes(String incoming) throws Exception {
...
-		log.debug("Trying to set value to " + propVal + " when propName is " + propName + " and context is "
-		        + n.getTextContent());
-		
+	public static void setProperty(Object o, Field field, String xmlFieldName, String xmlFieldValue, String xmlFieldType) throws Exception {
+		Object propVal = null;
+		if (xmlFieldValue != null) {
+			propVal = SyncUtil.valForField(field, xmlFieldValue, xmlFieldType);
+		}
+		log.debug("Trying to set value to " + propVal + " when propName is " + xmlFieldName + " and content is " + xmlFieldValue);
 		if (propVal != null) {
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/sync/SyncUtil.java:222]: @cioan / @mogoodrich .  I'm curious about this pre-existing behavior.  Why are we only setting the property if it is non-null?  What about the case where an existing property is non-null, and we are updating it via sync to a null value?  We should add some tests to make sure that behaves correctly and that skipping null values like this is the correct behavior.

File: api/src/main/java/org/openmrs/module/sync/SyncUtil.java [hunk 2]
```diff
@@ -289,212 +302,202 @@ public static Object newObject(String className) throws Exception {
...
+		if (field != null) {
+			classType = getFieldType(field);
+			if (classType == null) {
+				log.debug("type for " + field.getName() + " unable to be determined from field reflection");
+			}
+		}
+		else {
+			log.debug("field is null, trying to get class from xmlFieldType: " + xmlFieldType);
+		}
+		if (classType == null) {
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/sync/SyncUtil.java:373]: This block here is the main addition that got things working with PatientProgramAttributes.

File: api/src/main/java/org/openmrs/module/sync/SyncUtil.java [hunk 3]
```diff
@@ -289,212 +302,202 @@ public static Object newObject(String className) throws Exception {
...
+			}
+		}
+
+		// If none of the special cases handled above apply, convert the string value to an object of the field type
+		else {
+			o = convertStringToObject(xmlFieldValue, classType);
+		}
+
+		if (o == null) {
+			throw new IllegalArgumentException("Unable to convert " + xmlFieldValue + " into " + classType.getName());
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/sync/SyncUtil.java:498]: This is another area I'm curious about.  In the previous code, if the object was null here, the code just logged a debug message (see above) that no property value was able to be found.  I'm made things more strict, such that if this class can't identify a non-null value, then we fail hard with an exception.  This caused several unit tests to fail which had been getting away with using Dummy data previously, but fixing those to use property data was easy, and everything still passes.  I think throwing the Exception here is the right behavior.  Otherwise, sync could result in properties not getting set as intended, without anyone being aware of it.

File: api/src/main/java/org/openmrs/module/sync/SyncUtil.java (hunk 1 above)
Comment (mseaton) [reply api/src/main/java/org/openmrs/module/sync/SyncUtil.java:222]: Not sure I follow, this is a property, not an object.  So, for example, if you have an existing PatientIdentifiertype, with a non-null value for "format".  And then you update the PatientIdentifierType on the local server, setting "format" to null, does this successfully get set to null on the remote server?  I would be shocked if this was not happening, but just seeing that null values were ignored here in the code, it got me worried.  So we should write some tests around this to make sure that properties that are changed from a non-null value to a null value are correctly synced.

Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/sync/SyncUtil.java:222]: Yeah, agreed that I'd be surprised if that was not happened, because I figure we would have run into it as a problem before now, but definitely agree worth unit testing and also worth testing what happened on a sync testing dev environment.

**Comment by cioan** [comment]
@mseaton , I have added a couple more tests for editing an existing value and for creating a new patient program attribute of type Location. I could not figure out how to delete an existing PatientProgramAttribute. I could not find and OpenMRS API on how to delete this type of attribute.

File: api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java [hunk 4]
```diff
@@ -50,6 +52,81 @@ public String getInitialDataset() {
...
+
+				ProgramAttributeType programAttributeType = programService.getProgramAttributeType(2); // Health Center Location
+				Assertions.assertNotNull(programAttributeType);
+				PatientProgram pp = programService.getPatientProgram(1);
+				Assertions.assertNotNull(pp);
+				Location healthCenterLocation = Context.getLocationService().getLocation("Someplace");
+				Assertions.assertNotNull(healthCenterLocation);
+				PatientProgramAttribute attr = new PatientProgramAttribute();
+				attr.setAttributeType(programAttributeType);
+				attr.setValueReferenceInternal(healthCenterLocation.getUuid());
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java:71]: You should probably instead do `attr.setValue(healthCenterLocation)`, as this is the way the API would normally be used

File: api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java [hunk 5]
```diff
@@ -50,6 +52,81 @@ public String getInitialDataset() {
...
+			@Override
+			public void changedBeingApplied(List<SyncRecord> syncRecords, Record record) throws Exception {
+				super.changedBeingApplied(syncRecords, record);
+			}
+
+			public void runOnParent() throws Exception {
+				PatientProgram pp = programService.getPatientProgram(1);
+				Set<PatientProgramAttribute> attributes = pp.getAttributes();
+				Assertions.assertEquals(1, attributes.size());
+				Location location = Context.getLocationService().getLocationByUuid(attributes.iterator().next().getValueReference());
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java:88]: Again, here you probably want to get Location as `(Location) attributes().iterator().next().getValue()`

**Comment by mseaton** [comment]
Thanks @cioan - looks good.  Made a few comments just how we'd get/set the attribute values.  For deleting, I'd expect that if you remove an attribute from the program then it is deleted?  Can you test deleting out by seeing what happens if you remove an existing attribute from a patient program and save it.

File: api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java [hunk 6]
```diff
@@ -127,7 +128,46 @@ public void runOnParent() throws Exception {
...
+				List<PatientProgramAttribute> activeAttributes = pp.getActiveAttributes(programAttributeType);
+				Assertions.assertEquals(1, activeAttributes.size());
+				PatientProgramAttribute attr = activeAttributes.get(0);
+				Assertions.assertEquals(attr.getValue(), "transfer-in");
+				//activeAttributes.remove(0); -- it does not remove the attribute
+				//pp.getAttributes().remove(0);
+				Set<PatientProgramAttribute> ppAttributes = pp.getAttributes();
+				Iterator<PatientProgramAttribute> iterator = ppAttributes.iterator();
+				while(iterator.hasNext()){
+					iterator.remove();
```
Comment (cioan) [inline api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java:151]: @mseaton , I have made a few different unsuccessful attempts here to remove a patient_program_attribute. I am not sure if my syntax is not correct, or the OpenMRS API does not yet support this method? This last attempt that is not commented out, iterator.remove() throws IllegalStateException. Can you please have a look at this code, and let me know if there is a better way to do this? Thanks!

File: api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java [hunk 7]
```diff
@@ -143,13 +144,7 @@ public void runOnChild() throws Exception {
...
 				Assertions.assertEquals(attr.getValue(), "transfer-in");
-				//activeAttributes.remove(0); -- it does not remove the attribute
-				//pp.getAttributes().remove(0);
-				Set<PatientProgramAttribute> ppAttributes = pp.getAttributes();
-				Iterator<PatientProgramAttribute> iterator = ppAttributes.iterator();
-				while(iterator.hasNext()){
-					iterator.remove();
-				}
+				pp.getAttributes().remove(attr);
 				Assertions.assertEquals(0, pp.getAttributes().size());
```
[the remark this answers is not in the document]
Comment (cioan) [reply api/src/test/java/org/openmrs/module/sync/SyncProgramAttributeTest.java:148]: Thanks @mseaton ! This works great. The attribute is removed and synced successfully to the parent. Can we merge this PR now? Thanks!

File: api/src/main/java/org/openmrs/module/sync/SyncUtil.java (hunk 1 above)
Comment (mseaton) [reply api/src/main/java/org/openmrs/module/sync/SyncUtil.java:222]: @cioan - can you add a unit test for what I indicated above?  Basically, test a case where you are changing a property value of an existing object from non-null to null on the child, and then sync the object to the parent.  On the parent, this property should be set to null.

File: api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java [hunk 8]
```diff
@@ -74,7 +75,31 @@ public void runOnParent() throws Exception {
...
+				ps.savePerson(person);
+				Assertions.assertNull(person.getBirthdate());
+			}
+
+			public void runOnParent() throws Exception {
+				PersonService ps = Context.getPersonService();
+				// test to make sure that the null birthdate value has been synced
+				Person person = ps.getPerson(3);
+				Assertions.assertNull(person.getBirthdate());
+			}
```
Comment (cioan) [inline api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java:99]: @mseaton , you were right, if we change an object property to null it does not get synced. Here is a simple unit test case that confirms this suspicion. Since you refactored the sync code above, and you suspected this was a problem, do you mind fixing this problem with null values too? Thanks!

File: api/src/main/java/org/openmrs/module/sync/SyncUtil.java (hunk 1 above)
Comment (mseaton) [reply api/src/main/java/org/openmrs/module/sync/SyncUtil.java:222]: Follow-up on this conversation, I have created a new ticket for this issue, which does seem like a bug, but which is pre-existing and not related to this PR / change:  https://issues.openmrs.org/browse/SYNC-357

File: api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java (hunk 8 above)
Comment (mseaton) [reply api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java]: @cioan - it was not a simple fix, and a pre-existing issue, so I am creating a new ticket for us to address that here:  https://issues.openmrs.org/browse/SYNC-357

Comment (cioan) [reply api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java]: Thanks @mseaton ! This makes sense to me.

Comment (cioan) [reply api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java]: @mseaton , please let me know if you want me to merge this PR. Thanks!

Comment (mseaton) [reply api/src/test/java/org/openmrs/module/sync/SyncPersonTest.java]: Already merged @cioan thanks
